### DISCUSSION TITLE
Nuxt 3 vs Vitesse for complex a complex SPA

### DISCUSSION THREAD

**Discussion by binaryfire** [nuxt/nuxt#22968] [Questions]
Hi guys

We're developing a complex SPA similar to Figma. It's a pure client-side app that'll be delivered from a CDN (so no SSR/SSG) and is auth protected (so no SEO). The directory structure is modular and we're using Vite's Multi-Page App feature to split large features into separate pages.

We used Vitesse as a starting point and I'm trying to decide if it makes sense to migrate to Nuxt. A lot of people are saying "just use Nuxt for everything" but the benefits aren't clear to me in this case... Could anyone offer some guidance? I like how seamlessly integrated Nuxt's features are, but it seems like Vitesse might be more flexible for an "unusual" SPA. Am I missing something?

**Comment by Hebilicious** [comment]
@binaryfire There's a few open source complex nuxt SPA out there such as https://github.com/elk-zone/elk, I would take a look and see how you like it.

From the top of my head Nuxt will give you server features, better devtools and the module system, that is super neat for 3rd party module and for your private internal module too... 
The server features, even if you are not using SSR, are still useful depending on your app. If you are not using any of these, you might be interested in the Nuxt experimental features such as ViewTransitions etc, that you'd have to manually implement with Vitesse.

@antfu who created Vitesse is on the Nuxt team, maybe he will have more insight.

**Comment by binaryfire** [comment]
Thanks @Hebilicious

It's a pure SPA with a separate backend so I don't need any of the server stuff. I'm planning to deliver it from a regular CDN (no edge compute).

The Nuxt devtools are being ported at the moment too: https://github.com/webfansplz/vite-plugin-vue-devtools. So I think that'll be similar in the end.

The modules are nice but a lot of them are server and website oriented. [unjs.io](https://unjs.io/) seems to cover a lot of the bases too.

On the flip side, there are things that seem harder with Nuxt vs a traditional SPA:
- MPA. We need to be able to build certain features as independent pages with separate index.html files. We're using this Vite/rollup feature to build several mini-apps using the same codebase.
- Modular architecture. We need separate directories for each feature This app would be too hard to maintain otherwise.

Would love to hear @antfu's thoughts!

**Reply by Hebilicious** [reply]
The vite devtools are great. The Nuxt ones can do different things due to a tighter framework integration, not that it should sway your decision anyways. 

For modules (and layers), you have a powerful way to extend the framework, whereas with vite + vue you'd have to write your own conventions/abstractions.

With layers and module, you should have the flexibility to do what you want in terms of modularity (there's also an upcoming feature that might be related https://github.com/nuxt/nuxt/issues/21635). For directory structure you there's several things that you can customise, such as https://deltener.com/blog/nuxt-enterprise-patterns-component-management/

Good luck for your decision, I was in the exact same position last year and stuck with Vitesse. If I had to make this decision this year I would choose Nuxt.

**Comment by danielroe** [comment]
You might find [this tweet](https://twitter.com/antfu7/status/1676903120452870146) interesting.

You can always set `ssr: false` for pure SPA with many of the Nuxt benefits.
